src/arch/x86_64/kernel/virtio.rs

```diff
@@ -0,0 +1,689 @@
+	let length: usize = pci::read_config(bus, device, virtiocapoffset + 12) as usize; // get offset_of!(virtio_pci_cap, length)
+	debug!(
+		"Found virtio config bar as 0x{:x}, offset 0x{:x}, length 0x{:x}",
+		bar, offset, length
+	);
+
+	if (adapter.base_sizes[bar] as usize) < offset + length {
+		error!(
+			"virtio config struct does not fit in bar! Aborting! 0x{:x} < 0x{:x}",
+			adapter.base_sizes[bar],
+			offset + length
+		);
+		return None;
+	}
+
+	// base_addresses from bar are IOBASE?
+	// TODO: do proper memmapped bars in pci.rs
+	let barword = pci::read_config(bus, device, pci::PCI_BAR0_REGISTER + ((bar as u32) << 2));
+	debug!("Found bar{} as 0x{:x}", bar, barword);
+	assert!(barword & 1 == 0, "Not an memory mapped bar!");
+
+	let bartype = (barword >> 1) & 0b11;
+	assert!(bartype == 2, "Not a 64 bit bar!");
+
+	let prefetchable = (barword >> 3) & 1;
+	assert!(prefetchable == 1, "Bar not prefetchable, but 64 bit!");
+
+	let barwordhigh = pci::read_config(
+		bus,
+		device,
+		pci::PCI_BAR0_REGISTER + (((bar + 1) as u32) << 2),
+	);
+	//let barbase = barwordhigh << 33; // creates general protection fault... only when shifting by >=32 though..
+	let barbase: usize = ((barwordhigh as usize) << 32) + (barword & 0xFFFF_FFF0) as usize;
+
+	debug!(
+		"Mapping bar {} at 0x{:x} with length 0x{:x}",
+		bar, barbase, length
+	);
+	// corrosponding setup in eg Qemu @ https://github.com/qemu/qemu/blob/master/hw/virtio/virtio-pci.c#L1590 (virtio_pci_device_plugged)
+	// map 1 page (0x1000?) TODO: map "length"!
+	// for virtio-fs we are "lucky" and each cap is exactly one page (contiguous, but we dont care, map each on its own)
+	let membase = barbase as *mut u8;
+	let capbase = unsafe { membase.offset(offset as isize) as usize };
+	let virtualcapaddr = virtualmem::allocate(BasePageSize::SIZE).unwrap();
+	let mut flags = paging::PageTableEntryFlags::empty();
+	flags.device().writable().execute_disable();
+	paging::map::<BasePageSize>(virtualcapaddr, capbase, 1, flags);
+
+	/*
+		let mut slice = unsafe {
+				core::slice::from_raw_parts_mut(membase.offset(offset as isize), length)
+		};
+		info!("{:?}", slice);
+	*/
+	if virtiocaptype == VIRTIO_PCI_CAP_NOTIFY_CFG {
+		let notify_off_multiplier: u32 = pci::read_config(bus, device, virtiocapoffset + 16); // get offset_of!(virtio_pci_notify_cap, notify_off_multiplier)
+		Some((virtualcapaddr, notify_off_multiplier))
+	} else {
+		Some((virtualcapaddr, 0))
+	}
+}
+
+pub fn init_virtio_device(adapter: pci::PciAdapter) {
+	// TODO: 2.3.1: Loop until get_config_generation static, since it might change mid-read
+
+	if adapter.device_id <= 0x103F {
+		// Legacy device, skip
+		info!("Legacy Virtio device, skipping!");
+		return;
+	}
+	let virtio_device_id = adapter.device_id - 0x1040;
+
+	let drv = match virtio_device_id {
+		0x1a => {
+			info!("Found Virtio-FS device!");
+			// TODO: proper error handling on driver creation fail
+			virtio_fs::create_virtiofs_driver(adapter).unwrap()
+		}
```

Comment:
Do you have to check class_id and subclass_id? see [https://wiki.osdev.org/Virtio](https://wiki.osdev.org/Virtio)

Reply:
vendor_id and device_id are currently checked in pci.rs::init_drivers(), so init_virtio_device() is only called with the correct adapters. Since there currently is only one driver, I opted for this simple approach rather than the 'loop through all drivers' linux does. Another check here wouldn't hurt if this were to ever change though.

Reply:
As fas as I know, you are able to identify with the class_id the device type (block, network, etc.)

Reply:
I am not entirely sure what you mean. Class and subclass id from register 02 (https://wiki.osdev.org/PCI) are not mentioned in the [virtio spec](http://docs.oasis-open.org/virtio/virtio/v1.0/cs04/virtio-v1.0-cs04.html#x1-660001) as far as I can see. If you mean the Subsystem-ID and Subsystem-Vendor-ID, these are optional:

> The PCI Subsystem Vendor ID and the PCI Subsystem Device ID MAY reflect the PCI Vendor and Device ID of the environment (for informational purposes by the driver). 

I differentiate between the device types via the PCI Device ID (from register 0), which is 0x1040+virtio_id. This id is 0x1a for virtio-fs.

I suppose osdev mentiones the subsystem ID, since 
> Transitional devices MUST have the PCI Subsystem Device ID matching the Virtio Device ID, ...